scripts/clean-shrinkwrap.js

```diff
@@ -15,29 +15,222 @@
+ */
+const getDependencyTree = (shrinkwrap, shrinkwrapPath) => {
+  const dependencies = getTopLevelDependenciesFromShrinkwrapObject(shrinkwrapPath);
+
+  if (_.isEmpty(dependencies)) {
+    return [];
+  }
+
+  const object = getShrinkwrapDependencyObject(shrinkwrap, shrinkwrapPath);
+  const result = _.map(dependencies, (version, name) => {
+    const dependencyPath = _.has(object.dependencies, name) ? _.concat(shrinkwrapPath, [ name ]) : [ name ];
+    return getPrettyShrinkwrapDependencyObject(shrinkwrap, dependencyPath);
   });
+
+  return _.concat(result, _.flatMapDeep(result, (dependency) => {
+    return getDependencyTree(shrinkwrap, dependency.path);
+  }));
+};
+
+/**
+ * @summary Remove certain dependencies from a shrinkwrap object
+ * @function
+ * @private
+ *
+ * @description
+ * A shrinkwrap object is a recursive data structure, that apart
+ * from some extra metadata, has the following structure:
+ *
+ * {
+ *   ...
+ *   "dependencies": {
+ *     "<dependency_name>": <recursive definition>,
+ *     "<dependency_name>": <recursive definition>,
+ *     "<dependency_name>": <recursive definition>,
+ *     ...
+ *   }
+ * }
+ *
+ * The purpose of this function is to remove certain dependencies
+ * that match a blacklist. In order to do so, we start from the top
+ * level object, remove the blacklisted dependencies, and recurse
+ * if possible.
+ *
+ * @param {Object} shrinkwrap - the shrinkwrap file contents
+ * @param {String[]} blacklist - dependency blacklist
+ * @param {Object} qualifiers - qualifiers
+ * @param {Boolean} qualifiers.development - development dependency
+ * @param {Boolean} qualifiers.optional - optional dependency
+ * @returns {Object} filtered shrinkwrap object
+ *
+ * @example
+ * const shrinkwrapFile = require('./npm-shrinkwrap.json');
+ * const dependencyTree = getDependencyTree(shrinkwrapFile, [ 'drivelist' ]);
+ * const filteredShrinkwrap = removeDependencies(shrinkwrapFile, dependencyTree, {
+ *   optional: true,
+ *   development: true
+ * });
+ */
+const removeDependencies = (shrinkwrap, blacklist, qualifiers) => {
+  if (!_.isEmpty(shrinkwrap.dependencies)) {
+    shrinkwrap.dependencies = _.chain(shrinkwrap.dependencies)
+      .omitBy((dependency, name) => {
+        return _.every([
+          _.find(blacklist, {
+            name,
+            version: dependency.version
+          }),
+          dependency.dev === qualifiers.development,
+          dependency.optional === qualifiers.optional
+        ]);
+      })
+      .mapValues((dependency) => {
+        return removeDependencies(dependency, blacklist, qualifiers);
+      })
+      .value();
+  }
+
+  return shrinkwrap;
+};
+
+const blacklist = _.reduce(platformSpecificDependencies, (accumulator, dependencyPath) => {
+  return _.compact(_.concat(accumulator, [
+    getPrettyShrinkwrapDependencyObject(shrinkwrapFile, dependencyPath)
+  ], getDependencyTree(shrinkwrapFile, dependencyPath)));
+}, []);
```

Comment:
Haha, what this does is basically taking every dependency in the black list, and for each of them, get an array of its dependencies + the dependency itself (hence `_.concat`). We then flatten the whole thing as we go by using a fold.

I guess we can rename the internal bit to something like "get dependency tree AND the dependency itself". Maybe that could be just `getTree()`?